README.md
[![Pipeline](https://gitlab.com/bolderflight/software/pwm/badges/main/pipeline.svg)](https://gitlab.com/bolderflight/software/pwm/) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

![Bolder Flight Systems Logo](img/logo-words_75.png) &nbsp; &nbsp; ![Arduino Logo](img/arduino_logo_75.png)

# Pwm
This library sends commands to PWM servos and is compatible with Arduino and CMake build systems. 
   * [License](LICENSE.md)
   * [Changelog](CHANGELOG.md)
   * [Contributing guide](CONTRIBUTING.md)

# Description
Pulse Width Modulation (PWM) is a protocol for receivers to send commands to servos. Each servo is sent a high pulse with the width of the pulse dictating the commanded position. Traditionally, PWM commands are sent at a 50 Hz interval with pulse widths varying between 1000 us and 2000 us. Some servos and Electronic Speed Controllers (ESCs) support higher frequencies, which typically requires a shorter pulse width range. PWM is the most common protocol for sending servo and ESC commands, but they require a PWM capable pin for each servo being controlled.

# Usage
This library presents a similar interface as our [SBUS library](https://github.com/bolderflight/sbus/) and uses [analogWrite](https://www.pjrc.com/teensy/td_pulse.html) to generate the PWM pulses. *analogWrite* **does not** have a well defined interface across Arduino devices and this library is not anticipated to work with boards other than Teensy 3.x, 4.x, and LC. Compared to the [Servo library](https://www.arduino.cc/reference/en/libraries/servo/), *analogWrite* sends all the pulses at the same rising pulse, whereas *Servo* sends them one after another, which causes an undesirable and variable latency, especially in later channels.

# Installation

## Arduino
Simply clone or download and extract the zipped library into your Arduino/libraries folder. The library is added as:

```C++
#include "pwm.h"
```

An example is located in *examples/arduino/pwm_example/pwm_example.ino*. This library is tested with Teensy 3.x, 4.x, and LC devices and is **not** expected to work with other Arduino devices.

## CMake
CMake is used to build this library, which is exported as a library target called *pwm*. The header is added as:

```C++
#include "pwm.h"
```

The library can be also be compiled stand-alone using the CMake idiom of creating a *build* directory and then, from within that directory issuing:

```
cmake .. -DMCU=MK66FX1M0
make
```

This will build the library and example executable called *pwm_example*. The example executable source file is located at *examples/cmake/pwm_example.cc*. Notice that the *cmake* command includes a define specifying the microcontroller the code is being compiled for. This is required to correctly configure the code, CPU frequency, and compile/linker options. The available MCUs are:
   * MK20DX128
   * MK20DX256
   * MK64FX512
   * MK66FX1M0
   * MKL26Z64
   * IMXRT1062_T40
   * IMXRT1062_T41

These are known to work with the same packages used in Teensy products. Also switching the packages is known to work well, as long as it's only a package change.

The *pwm_example* target creates an executable for communicating with PWM servos. This target also has a *_hex* for creating the hex file and an *_upload* for using the [Teensy CLI Uploader](https://www.pjrc.com/teensy/loader_cli.html) to flash the Teensy. Please note that the CMake build tooling is expected to be run under Linux or WSL, instructions for setting up your build environment can be found in our [build-tools repo](https://github.com/bolderflight/build-tools). 

# Namespaces
This library is within the namespace *bfs*

# PwmTx

**PwmTx(const std::array<int8_t, N> &pins)** This class is templated with the number of PWM pins and takes an array of pin numbers as input.

```C++
std::array<int8_t, 6> pins = {21, 22, 23, 2, 3, 4};
bfs::PwmTx<pins.size()> pwm(pins);
```

An optional template parameter is the PWM resolution.

```C++
/* 12 bit PWM resolution */
std::array<int8_t, 6> pins = {21, 22, 23, 2, 3, 4};
bfs::PwmTx<pins.size(), 12> pwm(pins);
```

By default the resolution is 16 bits.

**void Begin()** Initializes the PWM pins at the default frequency of 50 Hz.

```C++
pwm.Begin();
```

**void Begin(const float freq)** Initializes the PWM pins at the given frequency in Hz. Note that each *PwmTx* object only supports a single frequency. PWM pins are grouped by timer, changing the frequency of any pin in the group changes the frequency for all pins in the group. Refer to the [analogWrite](https://www.pjrc.com/teensy/td_pulse.html) for the groupings.

**void Write()** Updates the PWM commands sent to the servos.

```C++
pwm.Write();
```

**static constexpr int8_t NUM_CH()** A constant defining the number of PWM channels, useful for defining arrays to write the data from.

**void ch(const std::array<int16_t, N> &cmds)** Sets the channel data to be transmitted, given an array of PWM commands.

```C++
std::array<int16_t, 6> cmd = {1000, 1200, 1300, 1400, 1500, 1600};
pwm.ch(cmd);
```

**std::array<int16_t, N> ch()** Returns the channel data to be transmitted.

```C++
std::array<int16_t, 6> cmd = pwm.ch();
```


CHANGELOG.md
# Changelog

## v6.0.2
- Bug fix with respect to setting analogWriteResolution

## v6.0.1
- size_t to std::size_t

## v6.0.0
- Updated to work with Arduino and CMake

## v5.1.0
- Updated to use mcu-support repo

## v5.0.0
- Moving away from Effector interface

## v4.0.4
- Update for effector v6.1.3, which adds a check to see whether an effector has been configured.

## v4.0.3
- Update for effector v6.1.0

## v4.0.2
- Pulling in effector v6.0.2

## v4.0.1
- Fixed constness of std::span

## v4.0.0
- Updated to conform to the Effector interface

## v3.0.1
- Updated pwm pin numbers to int8_t

## v3.0.0
- Updated to use a vector for entering the pin numbers.

## v2.0.0
- Updated to namespace *bfs*
- Renamed *Pwm* to *PwmTx* in case we write a *PwmRx* library in the future

## v1.0.5
- Updated for core v2.0.4

## v1.0.4
- Updated for core v2.0.3

## v1.0.3
- Baselined on core v1.0.3

## v1.0.2
- Updated CONTRIBUTING
- Updated *flash_mcu.cmake* to use local loader for Linux
- Updated *fetch_content* to use https instead of ssh

## v1.0.1
- Updated license to MIT
- Specified versions for dependencies

## v1.0.0
- Initial commit


src/pwm.h
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2021 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#ifndef SRC_PWM_H_
#define SRC_PWM_H_

#if defined(ARDUINO)
#include <Arduino.h>
#else
#include "core/core.h"
#endif
#include <cstddef>
#include <cstdint>
#include <cmath>
#include <array>

namespace bfs {

template<int8_t N, int8_t RES = 16>
class PwmTx {
 private:
  /* Pin numbers */
  std::array<int8_t, N> pins_;
  /* PWM resolution */
  static constexpr int8_t RES_ = RES;
  /* PWM bits */
  static constexpr float MAX_VAL_ =
    std::pow(2.0f, static_cast<float>(RES_)) - 1.0f;
  /* PWM frequency */
  float pwm_freq_hz_ = 50;
  /* PWM period */
  float pwm_period_us_;
  /* TX data */
  std::array<int16_t, N> ch_;

 public:
  explicit PwmTx(const std::array<int8_t, N> &pins) : pins_(pins) {}
  void Begin() {
    Begin(pwm_freq_hz_);
    analogWriteResolution(RES_);
  }
  void Begin(const float freq) {
    pwm_period_us_ = 1.0f / freq * 1000000.0f;
    /* Set the period and frequency */
    for (std::size_t i = 0; i < N; i++) {
      analogWriteFrequency(pins_[i], freq);
    }
  }
  void Write() {
    for (std::size_t i = 0; i < N; i++) {
      analogWrite(pins_[i], static_cast<float>(ch_[i]) /
                  pwm_period_us_ * MAX_VAL_);
    }
  }
  static constexpr int8_t NUM_CH() {return N;}
  inline void ch(const std::array<int16_t, N> &cmds) {ch_ = cmds;}
  inline std::array<int16_t, N> ch() const {return ch_;}
};

}  // namespace bfs

#endif  // SRC_PWM_H_


examples/cmake/pwm_example.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2021 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "pwm.h"

/* PWM object */
std::array<int8_t, 6> pins = {21, 22, 23, 2, 3, 4};
bfs::PwmTx<pins.size()> pwm(pins);

int main() {
  /* Serial to display data */
  Serial.begin(115200);
  while (!Serial) {}
  pwm.Begin();
  /* Issue a command */
  std::array<int16_t, 6> cmd = {1000, 1200, 1300, 1400, 1500, 1600};
  pwm.ch(cmd);
  pwm.Write();
  while (1) {}
}
